stdlib/universe/fill.go

```diff
@@ -204,59 +214,63 @@ func (t *fillTransformation) Process(id execute.DatasetID, tbl flux.Table) error
 		}
 	}
 
-	builder, created := t.cache.TableBuilder(key)
-	if created {
-		if err := execute.AddTableCols(tbl, builder); err != nil {
-			return err
-		}
-	}
-	idx := execute.ColIdx(t.spec.Column, builder.Cols())
-	if idx < 0 {
-		return errors.Newf(codes.FailedPrecondition, "fill column not found: %s", t.spec.Column)
-	}
-
 	prevNonNull := t.spec.Value
 	if !t.spec.UsePrevious {
-		if builder.Cols()[idx].Type != flux.ColumnType(prevNonNull.Type()) {
-			return errors.Newf(codes.FailedPrecondition, "fill column type mismatch: %s/%s", builder.Cols()[idx].Type.String(), flux.ColumnType(prevNonNull.Type()).String())
+		if tbl.Cols()[idx].Type != flux.ColumnType(prevNonNull.Type()) {
+			return errors.Newf(codes.FailedPrecondition, "fill column type mismatch: %s/%s", tbl.Cols()[idx].Type.String(), flux.ColumnType(prevNonNull.Type()).String())
 		}
 	}
-	return tbl.Do(func(cr flux.ColReader) error {
-		for j := range cr.Cols() {
-			if j == idx {
-				continue
-			}
-			if err := execute.AppendCol(j, j, cr, builder); err != nil {
-				return err
-			}
-		}
-		// Set new value
-		l := cr.Len()
 
-		if l > 0 {
-			if t.spec.UsePrevious {
-				prevNonNull = execute.ValueForRow(cr, 0, idx)
+	table, err := table.StreamWithContext(t.ctx, key, tbl.Cols(), func(ctx context.Context, w *table.StreamWriter) error {
+		return tbl.Do(func(cr flux.ColReader) error {
+			l := cr.Len()
+			if l == 0 {
+				return nil
 			}
-
-			for i := 0; i < l; i++ {
-				v := execute.ValueForRow(cr, i, idx)
-				if v.IsNull() {
-					if err := builder.AppendValue(idx, prevNonNull); err != nil {
-						return err
-					}
+			vs := make([]array.Interface, len(w.Cols()))
+			for j, col := range w.Cols() {
+				if j != idx {
+					vs[j] = table.Values(cr, j)
+					vs[j].Retain()
 				} else {
-					if err := builder.AppendValue(idx, v); err != nil {
-						return err
-					}
 					if t.spec.UsePrevious {
-						prevNonNull = v
+						prevNonNull = execute.ValueForRow(cr, 0, idx)
+					}
+					b := arrow.NewBuilder(col.Type, t.alloc)
+					b.Reserve(l)
```

Comment:
The `Reserve` method is a performance trap in Arrow for whatever reason. Use `Resize` instead. This is because `Reserve` will allocate an array that fits `l` elements and it will overallocate to the nearest multiple of 8. This is fine, but it combines with a rather negative behavior when `NewArray()` is called. When `NewArray()` is called, it will attempt to resize the allocated array to the smallest necessary space to fit the array. This results in a call to `Reallocate`. In C implementations, this probably results in a call to `realloc` which is probably quick when making the array size smaller and probably helps with keeping memory performant. In Go, there is no access to `realloc` from pure-Go so the reallocate is implemented in allocating a new array and copying the values which creates further churn on the garbage collector.

In the Go implementation of Arrow, if you know the exact size of the array, always use `Resize()`. There's not a lot of good reasons to use `Reserve()` in the Go implementation.